vmaware: init at 2.4.1

ners opened the pull request:
<img src="https://github.com/kernelwernel/VMAware/raw/main/assets/banner.jpg">

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ners committed:
vmaware: init at 2.4.1

ethancedwards8 commented on pkgs/by-name/vm/vmaware/package.nix:
```diff
@@ -0,0 +1,59 @@
+
+  meta = {
+    description = "VM detection library and tool";
+    homepage = "https://github.com/kernelwernel/VMAware";
```
```suggestion
    homepage = "https://github.com/kernelwernel/VMAware";
    changelog = "https://github.com/kernelwernel/VMAware/releases/tag/v${version}";
```

ethancedwards8 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 390431`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>vmaware</li>
  </ul>
</details>

ethancedwards8 commented on pkgs/by-name/vm/vmaware/package.nix:
```diff
@@ -0,0 +1,59 @@
+    wrapProgram $out/bin/vmaware \
+      --prefix PATH : "${lib.makeBinPath runtimeDeps}"
+  '';
+
```
add versionCheckHook

ethancedwards8 reviewed (approved):
Thanks!

ners commented:
This currently fails to build on ARM64, which has been fixed since v2.0. I'm drafting this until upstream cuts a new release.

kernelwernel commented:
The 2.1 will be released tomorrow for anybody reading this.

mattpolzin commented on pkgs/by-name/vm/vmaware/package.nix:
```diff
@@ -0,0 +1,67 @@
+    description = "VM detection library and tool";
+    homepage = "https://github.com/kernelwernel/VMAware";
+    changelog = "https://github.com/kernelwernel/VMAware/releases/tag/v${finalAttrs.version}";
+    license = with lib.licenses; [
+      gpl3
+      mit
+    ];
```
The licensing may be a bit too granular to represent simply as "one of MIT or GPL3."

The source repository describes the MIT licensed version as having fewer capabilities, but the Nixpkgs version will not be conditionally built with different capabilities depending on the license. Not sure the best option, but perhaps one option is to only list it as GPL3 here and add a code comment about why MIT was omitted.

mattpolzin replied:
I apologize, I somehow missed that the list form of licenses merely indicates _parts_ of the package are under each listed license. So that's appropriate here.

mattpolzin commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 390431`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>vmaware</li>
  </ul>
</details>

mattpolzin reviewed (changes requested):
Actually, I was able to reproduce the CI results of failure on x86_64-darwin so that should be marked as unsupported unless it can be fixed before merge.

Result of `nixpkgs-review pr 390431` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>vmaware</li>
  </ul>
</details>

kernelwernel commented:
Hi! This might sound a bit sudden but we've released a 2.1.1 that fixes a critical issue in the 2.1 release.

I would've made a fork and modified the .nix file myself to make this change quicker, but I'm not sure how or where the hash (`sha256-lnDxgSvElm8swSxVm+s7S9wz1Fy9Jb/Z+sLngowtnA0=`) of the release was generated.

mattpolzin commented:
Just so you know, there are a few ways to get that hash updated.

1. `nix-update`. In this case, the `nix-update` tool is able to find the new release so `nix run .#nix-update -- vmaware` does the trick (without flakes enabled, `nix-shell -p nix-update --run 'nix-update vmaware'`). This updates the `vmaware` derivation for you.
2. `nix-prefetch-git`. You can get the hash with `nix run .\#nix-prefetch-github -- --rev v2.1.1 kernelwernel vmaware`. Then you can plug it into the `vmaware` derivation yourself.
3. You can delete the hash (leave an empty string), build the derivation, and let the Nix error tell you what the hash should be. Most folks likely do this because it is so convenient.

ners commented:
@kernelwernel I've updated to 2.1.1, thanks! However with this release the version check fails; the program reports version 2.1 when run with the `--version` flag. I've removed the version check for now, but it would be nice if it was reported correctly.

mattpolzin commented:
@ners this is still failing to build for `x86_64-darwin`. can you mark it as unsupported for that platform (works for `aarch64-darwin`)? alternatively, maybe @kernelwernel has input RE supporting that platform.

kernelwernel commented:
> @kernelwernel I've updated to 2.1.1, thanks! However with this release the version check fails; the program reports version 2.1 when run with the `--version` flag. I've removed the version check for now, but it would be nice if it was reported correctly.

I’ll check why that’s happening but it should be an easy fix.

Unfortunately I’m in a different country for a university trip at the moment, and I can’t access the hotel wifi from my laptop for the past few days no matter what I do. I should be back in 2 or 3 days though, sorry.


> @ners this is still failing to build for `x86_64-darwin`. can you mark it as unsupported for that platform (works for `aarch64-darwin`)? alternatively, maybe @kernelwernel has input RE supporting that platform.

Strange, is it possible if I can have the error message(s)?

mattpolzin commented:
>  is it possible if I can have the error message(s)?

@kernelwernel absolutely!

```
In file included from /tmp/nix-build-vmaware-2.1.1.drv-0/source/src/cli.cpp:48:
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:196:9: warning: 'LINUX' macro redefined [-Wmacro-redefined]
  196 | #define LINUX 0
      |         ^
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/cli.cpp:32:13: note: previous definition is here
   32 |     #define LINUX 1
      |             ^
In file included from /tmp/nix-build-vmaware-2.1.1.drv-0/source/src/cli.cpp:48:
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:776:13: error: use of undeclared identifier '__cpuid_count'
  776 |             __cpuid_count(a_leaf, c_leaf, a, b, c, d);
      |             ^
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:797:13: error: use of undeclared identifier '__cpuid_count'
  797 |             __cpuid_count(a_leaf, c_leaf, x[0], x[1], x[2], x[3]);
      |             ^
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:1635:27: warning: unused variable 'GB' [-Wunused-variable]
 1635 |             constexpr u64 GB = (static_cast<u64>(1024) * 1024 * 1024);  // Size of 1 GB in bytes
      |                           ^~
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:8192:13: error: use of undeclared identifier '__cpuid'
 8192 |             __cpuid(0, eax, ebx, ecx, edx);
      |             ^
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:8216:13: error: unknown type name 'thread_affinity_policy_data_t'
 8216 |             thread_affinity_policy_data_t policy = { 1 };
      |             ^
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:8218:17: error: use of undeclared identifier 'THREAD_AFFINITY_POLICY'
 8218 |                 THREAD_AFFINITY_POLICY,
      |                 ^
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:8219:18: error: use of undeclared identifier 'thread_policy_t'
 8219 |                 (thread_policy_t)&policy, 1);
      |                  ^
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:8228:17: error: use of undeclared identifier '__cpuid'
 8228 |                 __cpuid(0, eax, ebx, ecx, edx);
      |                 ^
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:8239:9: error: use of undeclared identifier 'sched_setscheduler'
 8239 |         sched_setscheduler(0, SCHED_FIFO, &param);
      |         ^
/tmp/nix-build-vmaware-2.1.1.drv-0/source/src/vmaware.hpp:8251:13: error: use of undeclared identifier '__cpuid'
 8251 |             __cpuid(0, eax, ebx, ecx, edx);
      |             ^
2 warnings and 9 errors generated.
ninja: build stopped: subcommand failed.
```

kernelwernel commented:
@mattpolzin my bad, I didn't see the message. The lib is now a 2.4.1 instead of the 2.1.1, and I assume this should be fixed by now.

mattpolzin commented:
@kernelwernel different error message but still fails to build for `x86_64-darwin`. As I suggested above, marking that platform as unsupported is always an option to get this package merged into nixpkgs. You could put up another PR to mark it supported again in the future.

kernelwernel commented:
> Works for me. I am happy to build on my x86 machine and provide error messages since waiting for queued CI builds on that platform can take quite a while.

I've had a few MacOS guys use the release and I've never encountered the error you got, not sure why.

But anyways, I'm 99% the issue is fixed [here](https://github.com/kernelwernel/VMAware/tree/f1d0e38db68319b238140d1b7657b1ed804812fb). 

I was also thinking of releasing a 2.4.2 which is pretty much ready at this point. Would you prefer having the upstream 2.4.1 or wait for the 2.4.2? If it's the latter, I can release it within 24 hours.
